python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,716 @@
+                                              priority, 
+                                              group = self._bg_tasks_group,
+                                              task_args = task_args,
+                                              task_kwargs = task_kwargs)
+        return str(task_id)
+
+
+    def request_thumbnail(self, url, entity_type, entity_id, field, load_image=False):
+        """
+        Downloads a thumbnail form Shotgun asyncronously or returns the cached thumbnail 
+        if found.  Optionally loads the thumbnail into a QImage.
+
+        :param url:         The thumbnail url string that is associated with this thumbnail. This is
+                            the field value as returned by a Shotgun query.
+        :param entity_type: Shotgun entity type with which the thumb is associated.
+        :param entity_id:   Shotgun entity id with which the thumb is associated.
+        :param field:       Thumbnail field. Normally 'image' but could also for example be a deep 
+                            link field such as 'sg_sequence.Sequence.image'
+        :param load_image:  If set to True, the return data structure will contain a QImage object 
+                            with the image data loaded.
+
+        :returns:           A unique task id representing this request.
+        """
+        if not self._task_manager:
+            return
+
+        # always add check for thumbnail already downloaded:
+        check_task_id = self._task_manager.add_task(self._task_check_thumbnail,
+                                                    priority = ShotgunDataRetriever._CHECK_THUMB_PRIORITY,
+                                                    group = self._bg_tasks_group,
+                                                    task_kwargs = {"url":url, 
+                                                                   "load_image":load_image})
+
+        # add download thumbnail task
+        dl_task_id = self._task_manager.add_task(self._task_download_thumbnail,
+                                                 upstream_task_ids = [check_task_id],
+                                                 priority = ShotgunDataRetriever._DOWNLOAD_THUMB_PRIORITY,
+                                                 group = self._bg_tasks_group,
+                                                 task_kwargs = {"url":url,
+                                                                "entity_type":entity_type, 
+                                                                "entity_id":entity_id,
+                                                                "field":field,
+                                                                "load_image":load_image})
+
+        # all results for requesting a thumbnail should be returned with the same id so use
+        # a mapping to track the 'primary' task id:
+        self._thumb_task_id_map[dl_task_id] = check_task_id
+
+        return str(check_task_id)
```

Comment:
For this merge I'm going to leave the casting from int to string and back but I agree that it would be better if this wasn't happening - it is purely for backwards compatibility where user code that uses the data retriever may break if an int is returned (I had this happening somewhere at some point I think).

The task ids are guaranteed to be unique within the scope of a single BackgroundTaskManager so the only scenario it would break would be if an instance subscribed to two different BackgroundTaskManager instances with the same slot and didn't validate the sender - not a  pattern that makes much sense.  Using GUIDS just seemed a little overkill and (imo) harder to read/scan output when debugging